File: benchmarks/spectral-norm/spectral_norm.py
#!/usr/bin/env python3
"""
The Computer Language Benchmarks Game
https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

spectral-norm benchmark in Python
"""

import sys
import math

def eval_A(i, j):
    return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)

def eval_A_times_u(u, Au):
    n = len(u)
    for i in range(n):
        Au[i] = 0
        for j in range(n):
            Au[i] += eval_A(i, j) * u[j]

def eval_At_times_u(u, Au):
    n = len(u)
    for i in range(n):
        Au[i] = 0
        for j in range(n):
            Au[i] += eval_A(j, i) * u[j]

def eval_AtA_times_u(u, AtAu):
    n = len(u)
    v = [0] * n
    eval_A_times_u(u, v)
    eval_At_times_u(v, AtAu)

def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100
    u = [1.0] * n
    v = [0.0] * n
    
    for i in range(10):
        eval_AtA_times_u(u, v)
        eval_AtA_times_u(v, u)
    
    vBv = vv = 0
    for i in range(n):
        vBv += u[i] * v[i]
        vv += v[i] * v[i]
    
    print(f"{math.sqrt(vBv / vv):.9f}")

if __name__ == "__main__":
    main()

File: benchmarks/spectral-norm/spectral_norm.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// spectral-norm benchmark in Rust

use std::env;

fn eval_a(i: usize, j: usize) -> f64 {
    1.0 / ((i + j) * (i + j + 1) / 2 + i + 1) as f64
}

fn eval_a_times_u(u: &[f64], v: &mut [f64]) {
    let n = u.len();
    for i in 0..n {
        v[i] = 0.0;
        for j in 0..n {
            v[i] += eval_a(i, j) * u[j];
        }
    }
}

fn eval_at_times_u(u: &[f64], v: &mut [f64]) {
    let n = u.len();
    for i in 0..n {
        v[i] = 0.0;
        for j in 0..n {
            v[i] += eval_a(j, i) * u[j];
        }
    }
}

fn eval_ata_times_u(u: &[f64], v: &mut [f64], w: &mut [f64]) {
    eval_a_times_u(u, w);
    eval_at_times_u(w, v);
}

fn main() {
    let n = env::args()
        .nth(1)
        .and_then(|n| n.parse().ok())
        .unwrap_or(100);
    
    let mut u = vec![1.0; n];
    let mut v = vec![0.0; n];
    let mut w = vec![0.0; n];
    
    for _ in 0..10 {
        eval_ata_times_u(&u, &mut v, &mut w);
        eval_ata_times_u(&v, &mut u, &mut w);
    }
    
    let mut vbv = 0.0;
    let mut vv = 0.0;
    
    for i in 0..n {
        vbv += u[i] * v[i];
        vv += v[i] * v[i];
    }
    
    println!("{:.9}", (vbv / vv).sqrt());
}
